File: app/src/main/java/org/hackru/oneapp/hackru/LoginActivity.java
package org.hackru.oneapp.hackru;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import org.hackru.oneapp.hackru.api.model.AuthorizeRequest;
import org.hackru.oneapp.hackru.api.model.Login;
import org.hackru.oneapp.hackru.api.service.HackRUService;
import org.hackru.oneapp.hackru.utils.SharedPreferencesUtility;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class LoginActivity extends AppCompatActivity {
    private static final String TAG = "LoginActivity";

    @BindView(R.id.input_email) EditText _emailText;
    @BindView(R.id.input_password) EditText _passwordText;
    @BindView(R.id.btn_login) Button _loginButton;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        ButterKnife.bind(this);

        _loginButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                login();
            }
        });
    }

    public void login() {
        // Check for improper input
        if (!validate()) {
            _loginButton.setEnabled(true);
            return;
        }

        _loginButton.setEnabled(false);

        final ProgressDialog progressDialog = new ProgressDialog(LoginActivity.this,
                R.style.AppTheme_ProgressDialog);
        progressDialog.setIndeterminate(true);
        progressDialog.setMessage("Authenticating...");
        progressDialog.show();

        final String email = _emailText.getText().toString(); // Needs to be declared final so that we can use it in the overrided onResponse method below
        String password = _passwordText.getText().toString();

        // Authentication logic
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://m7cwj1fy7c.execute-api.us-west-2.amazonaws.com/mlhtest/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        HackRUService hackRUService = retrofit.create(HackRUService.class);
        hackRUService.authorize(new AuthorizeRequest(email, password)).enqueue(new Callback<Login>() {
            @Override
            public void onResponse(Call<Login> call, Response<Login> response) {
                Log.i(TAG, "Post submitted to API!");
                if(response.body().getStatusCode() == 200) {
                    String body = response.body().getBody();
                    String token = body.substring(body.indexOf("token")+9, body.indexOf(',')-1);
                    SharedPreferencesUtility.setAuthToken(LoginActivity.this, token);
                    SharedPreferencesUtility.setEmail(LoginActivity.this, email);
                    onLoginSuccess();
                } else {
                    onLoginFailed(true);
                }
                progressDialog.dismiss();
            }

            @Override
            public void onFailure(Call<Login> call, Throwable t) {
                Log.e(TAG, "Unable to submit post to API.");
                onLoginFailed(false);
                progressDialog.dismiss();
            }
        });


//        new android.os.Handler().postDelayed(
//                new Runnable() {
//                    public void run() {
//                        // On complete call either onLoginSuccess or onLoginFailed
//                        onLoginSuccess();
//                        // onLoginFailed();
//                        progressDialog.dismiss();
//                    }
//                }, 1000);
    }

    @Override
    public void onBackPressed() {
        // Disable going back to the MainActivity
        moveTaskToBack(true);
    }

    public void onLoginSuccess() {
        _loginButton.setEnabled(true);
        Intent loginActivityIntent = new Intent(this, MainActivity.class);
        startActivity(loginActivityIntent);
        finish();
    }

    public void onLoginFailed(Boolean connectionSuccess) {
        if(connectionSuccess) {
            Toast.makeText(getBaseContext(), "Incorrect username or password", Toast.LENGTH_LONG).show();
        } else {
            Toast.makeText(getBaseContext(), "Cannot reach server", Toast.LENGTH_LONG).show();
        }

        _loginButton.setEnabled(true);
    }

    public boolean validate() {
        boolean valid = true;

        String email = _emailText.getText().toString();
        String password = _passwordText.getText().toString();

        if (email.isEmpty() || !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            _emailText.setError("Enter a valid email address");
            valid = false;
        } else {
            _emailText.setError(null);
        }

        if (password.isEmpty() || password.length() < 4 || password.length() > 10) {
            _passwordText.setError("Between 4 and 10 alphanumeric characters");
            valid = false;
        } else {
            _passwordText.setError(null);
        }

        return valid;
    }
}

File: app/src/main/java/org/hackru/oneapp/hackru/ui/drawer/LoginActivity.kt
package org.hackru.oneapp.hackru.ui.drawer

import android.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.TextView
import android.widget.Toast
import com.google.gson.Gson
import kotlinx.android.synthetic.main.activity_login.*
import org.hackru.oneapp.hackru.HackRUApp
import org.hackru.oneapp.hackru.R
import org.hackru.oneapp.hackru.Utils
import org.hackru.oneapp.hackru.api.models.AuthorizeModel
import org.hackru.oneapp.hackru.api.models.ReadModel
import org.hackru.oneapp.hackru.api.services.LcsService
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException
import java.text.SimpleDateFormat
import javax.inject.Inject

class LoginActivity : AppCompatActivity() {
    val TAG = "LoginActivity"

    @Inject lateinit var lcsService: LcsService
    val gson = Gson()
    lateinit var loadingDialog: AlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        (application as HackRUApp).appComponent.inject(this)

        // Auto-fill the email of the last logged-in user
        val email = Utils.SharedPreferences.getEmail(this)
        input_email.setText(email, TextView.BufferType.EDITABLE)

        loadingDialog = AlertDialog.Builder(this)
                .setView(layoutInflater.inflate(R.layout.dialog_progress_circle, null))
                .setTitle("Authenticating...")
                .setCancelable(false)
                .create()

        button_login.setOnClickListener { login() }
    }

    // TODO: Make this gross code follow MVVM and app architecture guidelines
    fun login() {
        if(!inputIsValid()) return
        loadingDialog.show()
        val email = input_email.text.toString()
        val password = input_password.text.toString()
        val request = AuthorizeModel.Request(email, password)
        lcsService.authorize(request).enqueue(object : Callback<AuthorizeModel.Response> {
            override fun onResponse(call: Call<AuthorizeModel.Response>, response: Response<AuthorizeModel.Response>) {
                if(response.isSuccessful) {
                    val retrofitBody = response.body()
                    if(retrofitBody?.statusCode == 200) {
                        val body: String = retrofitBody.body
                        val auth: AuthorizeModel.Auth = gson.fromJson<AuthorizeModel.Body>(body, AuthorizeModel.Body::class.java).auth
                        Utils.SharedPreferences.setEmail(this@LoginActivity, email)
                        Utils.SharedPreferences.setAuthToken(this@LoginActivity, auth.token)
                        val logoutAt: Long = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSS")
                                .parse(auth.validUntil)
                                .time
                        Utils.SharedPreferences.setLogoutAt(this@LoginActivity, logoutAt)
                        retrieveNameAndCanScan(email, auth.token)
                    } else {
                        val errorMessage = response.body()?.body ?: "Network error. Please try again"
                        showToast(errorMessage, Toast.LENGTH_SHORT)
                        loadingDialog.dismiss()
                    }
                } else {
                    showToast("Network error. Please try again", Toast.LENGTH_SHORT)
                    loadingDialog.dismiss()
                }
            }

            override fun onFailure(call: Call<AuthorizeModel.Response>, t: Throwable) {
                if(t is IOException) {
                    showToast("Network error. Please try again", Toast.LENGTH_SHORT)
                    loadingDialog.dismiss()
                } else {
                    showToast("Serialization error. Please report this error to rnd@hackru.org", Toast.LENGTH_LONG)
                    loadingDialog.dismiss()
                }
            }
        })
    }

    fun retrieveNameAndCanScan(email: String, authToken: String) {
        val request = ReadModel.Request(email, authToken, ReadModel.Query(email))
        lcsService.read(request).enqueue(object : Callback<ReadModel.Response> {
            override fun onResponse(call: Call<ReadModel.Response>, response: Response<ReadModel.Response>) {
                if(response.isSuccessful) {
                    val retrofitBody = response.body()
                    if(retrofitBody?.statusCode == 200) {
                        val role: ReadModel.Role = retrofitBody.body[0].role
                        Utils.SharedPreferences.setCanScan(this@LoginActivity, role.director || role.organizer)
                        val firstName: String = retrofitBody.body[0].firstName
                        val lastName: String = retrofitBody.body[0].lastName
                        var fullName: String = "$firstName $lastName"
                        if(firstName.isEmpty() || lastName.isEmpty()) {
                            fullName = ""
                        }
                        Utils.SharedPreferences.setName(this@LoginActivity, fullName)
                        loadingDialog.dismiss()
                        finish()
                    } else {
                        showToast("Error reading user data. Please try again", Toast.LENGTH_SHORT)
                        loadingDialog.dismiss()
                    }
                } else {
                    showToast("Network error. Please try again", Toast.LENGTH_SHORT)
                    loadingDialog.dismiss()
                }
            }

            override fun onFailure(call: Call<ReadModel.Response>, t: Throwable) {
                if(t is IOException) {
                    showToast("Network error. Please try again", Toast.LENGTH_SHORT)
                    loadingDialog.dismiss()
                } else {
                    showToast("Serialization error. Please report this error to rnd@hackru.org", Toast.LENGTH_LONG)
                    loadingDialog.dismiss()
                }
            }
        })
    }

    fun showToast(message: String, duration: Int) {
        Toast.makeText(this, message, duration).show()
    }

    /**
     * Makes sure the provided email is not blank and is in email form. Also makes rue that the
     * provided password is not blank. If either the email or password do not meet these conditions,
     * then return false and show a warning dialog to the user.
     */
    fun inputIsValid(): Boolean {
        val email = input_email.text.toString()
        val password = input_password.text.toString()

        if(email.isEmpty() || !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            input_email.error = "Please enter a valid email address"
            return false
        } else {
            input_email.error = null
        }

        if(password.isEmpty()) {
            input_password.error = "Please enter a password"
            return false
        } else {
            input_password.error = null
        }

        return true
    }
}
